File: app/src/main/java/com/albert/fitness/pumpit/adapter/new_adapter/ExerciseAdapter.java
package com.albert.fitness.pumpit.adapter.new_adapter;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.databinding.DataBindingUtil;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.RecyclerView;

import com.albert.fitness.pumpit.model.Exercise;

import java.util.ArrayList;
import java.util.List;

import fitness.albert.com.pumpit.R;
import fitness.albert.com.pumpit.databinding.ExerciseListItemBinding;

public class ExerciseAdapter extends RecyclerView.Adapter<ExerciseAdapter.ExerciseViewHolder> {
    private OnItemClickListener listener;
    private ArrayList<Exercise> items = new ArrayList<>();

    @NonNull
    @Override
    public ExerciseViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int i) {

        ExerciseListItemBinding listItemBinding = DataBindingUtil.inflate(LayoutInflater.from(viewGroup.getContext()),
                R.layout.exercise_list_item, viewGroup, false);
        return new ExerciseViewHolder(listItemBinding);
    }

    @Override
    public void onBindViewHolder(@NonNull ExerciseViewHolder viewHolder, int i) {
        Exercise item = items.get(i);
        viewHolder.listItemBinding.setExercise(item);
    }

    @Override
    public int getItemCount() {
        return items.size();
    }

    public void updateList(List<Exercise> newList) {
        items = new ArrayList<>();
        items.addAll(newList);
        notifyDataSetChanged();
    }


    public void setItems(ArrayList<Exercise> newItems) {
        final DiffUtil.DiffResult result = DiffUtil.calculateDiff(new ExerciseDiffCallback(items, newItems), false);
        items = newItems;
        result.dispatchUpdatesTo(ExerciseAdapter.this);

    }

    class ExerciseViewHolder extends RecyclerView.ViewHolder {
        private ExerciseListItemBinding listItemBinding;

        public ExerciseViewHolder(@NonNull ExerciseListItemBinding listItemBinding) {
            super(listItemBinding.getRoot());
            this.listItemBinding = listItemBinding;
            listItemBinding.getRoot().setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    int clickedPosition = getAdapterPosition();

                    if (listener != null && clickedPosition != RecyclerView.NO_POSITION) {
                        listener.onItemClick(items.get(clickedPosition));
                    }
                }
            });

        }

    }

    public interface OnItemClickListener {
        void onItemClick(Exercise item);
    }

    public void setListener(OnItemClickListener listener) {
        this.listener = listener;
    }

    public class ExerciseDiffCallback extends DiffUtil.Callback {
        ArrayList<Exercise> oldItemsList;
        ArrayList<Exercise> newItemList;

        public ExerciseDiffCallback(ArrayList<Exercise> oldItemsList, ArrayList<Exercise> newItemList) {
            this.oldItemsList = oldItemsList;
            this.newItemList = newItemList;
        }

        @Override
        public int getOldListSize() {
            return oldItemsList == null ? 0 : oldItemsList.size();
        }

        @Override
        public int getNewListSize() {
            return newItemList == null ? 0 : newItemList.size();
        }

        @Override
        public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
            return oldItemsList.get(oldItemPosition).getExerciseId() == newItemList.get(newItemPosition).getExerciseId();
        }

        @Override
        public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
            return oldItemsList.get(oldItemPosition).equals(newItemList.get(newItemPosition));
        }

        @Nullable
        @Override
        public Object getChangePayload(int oldItemPosition, int newItemPosition) {
            return super.getChangePayload(oldItemPosition, newItemPosition);
        }
    }
}

File: app/src/main/java/com/albert/fitness/pumpit/adapter/new_adapter/ExerciseAdapter.kt
package com.albert.fitness.pumpit.adapterimport

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.AsyncListDiffer
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import com.albert.fitness.pumpit.helper.BitmapFromAssent
import com.albert.fitness.pumpit.model.Exercise
import com.bumptech.glide.Glide
import fitness.albert.com.pumpit.R
import kotlinx.android.synthetic.main.item_list_exercise.view.*

class ExerciseAdapter(private val interaction: Interaction? = null) :
        RecyclerView.Adapter<RecyclerView.ViewHolder>() {
    val DIFF_CALLBACK = object : DiffUtil.ItemCallback<Exercise>() {
        override fun areItemsTheSame(oldItem: Exercise, newItem: Exercise): Boolean {
            return  oldItem.exerciseId == newItem.exerciseId
        }

        override fun areContentsTheSame(oldItem: Exercise, newItem: Exercise): Boolean {
            return oldItem == newItem
        }
    }
    private val differ = AsyncListDiffer(this, DIFF_CALLBACK)
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        return ExerciseViewHolder(
                LayoutInflater.from(parent.context).inflate(
                        R.layout.item_list_exercise,
                        parent,
                        false
                ),
                interaction
        )
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        when (holder) {
            is ExerciseViewHolder -> {
                holder.bind(differ.currentList.get(position))
            }
        }
    }

    override fun getItemCount(): Int {
        return differ.currentList.size
    }

    fun submitList(list: List<Exercise>) {
        differ.submitList(list)
    }

    class ExerciseViewHolder
    constructor(
            itemView: View,
            private val interaction: Interaction?
    ) : RecyclerView.ViewHolder(itemView) {
        fun bind(item: Exercise) = with(itemView) {
            itemView.setOnClickListener {
                interaction?.onItemSelected(adapterPosition, item)
            }
            val bitmap = BitmapFromAssent.getBitmapFromAsset(itemView.context,item.imgName)
                        Glide.with(itemView.context)
                   .load(bitmap)
                  .into(itemView.exercise_img)

            itemView.tv_exercise_name.text = item.exerciseName
        }
    }

    interface Interaction {
        fun onItemSelected(position: Int, item: Exercise)
    }
}